### ISSUE TITLE
Highlighting is not applied correctly text is changed in code behind

### ISSUE THREAD

**Issue by guilhermeolisi** [AvaloniaUI/AvaloniaEdit#221]
Hi,

When I change part of the text in code behind, some times the highlighting is not applied correctly. It is intermittent, the same changes may highlighting correctly or not.

I made a test project (in file attached bellow). I'm using C# Grammar to highlighting. By clicking on the button, the text will be changed in code behind, the text in lines 3 and 4 will be commented out (inserting "//"). By clicking the same button again will be uncommented (removing "//") the same part of the text.

[AvaloniaEditTest.zip](https://github.com/AvaloniaUI/AvaloniaEdit/files/8389574/AvaloniaEditTest.zip)

The next images are a print screen of text after a few clicks, exemplifying the strange behavior:

Initial text:
![image](https://user-images.githubusercontent.com/25594740/161057832-2eeb76fa-978a-48fa-82ce-e35e4ee929cf.png)

Correct highlighting applied:
![image](https://user-images.githubusercontent.com/25594740/161058004-ae0e3ade-6be3-4386-bc1b-f42fc5ae83c5.png)

but sometimes the highlighting is applied wrong:

![image](https://user-images.githubusercontent.com/25594740/161058123-f5e933dd-a60b-4090-983a-c59685728d5d.png)

![image](https://user-images.githubusercontent.com/25594740/161058097-57e68ada-20c1-48a3-a515-c430f36afd6b.png)

![image](https://user-images.githubusercontent.com/25594740/161058163-3e3f2b7a-eac2-4b02-ac1e-25c5d8ed6e49.png)

**Comment by danipen** [comment]
Yeah, it's also reproduced when you paste text:
![syntax](https://user-images.githubusercontent.com/501613/161065331-574e56a9-1f33-45e2-9504-f2f9f29a0385.gif)

So it seems to be a bug in the code that processes the multiline editions. I'll try to take a look.

**Comment by danipen** [comment]
@guilhermeolisi please could you try if #222 fixes your issue? Thanks!

**Comment by guilhermeolisi** [comment]
hi @danipen,

I did the pull of AvaloniaEdit from GitHub with your modifications and changed the AvaloniaEdit.Demo to include my test button to change the text in code behind. The result was the same, sometimes the highlighting goes wrong. I recorded a Gif of my test:

![Test Avaloniaedit](https://user-images.githubusercontent.com/25594740/161320458-d6b0f6fc-46d3-48d2-b319-afd2e1b1644c.gif)

**Comment by guilhermeolisi** [comment]
Hi @danipen, did you manage to read my last message?


> hi @danipen,
> 
> I did the pull of AvaloniaEdit from GitHub with your modifications and changed the AvaloniaEdit.Demo to include my test button to change the text in code behind. The result was the same, sometimes the highlighting goes wrong. I recorded a Gif of my test:
> 
> ![Test Avaloniaedit](https://user-images.githubusercontent.com/25594740/161320458-d6b0f6fc-46d3-48d2-b319-afd2e1b1644c.gif)

**Comment by danipen** [comment]
Yes. I need to get some time to take a look into it. I'm quite busy now sorry.

**Comment by guilhermeolisi** [comment]
Ok, just to see if you've seen it, thanks

**Comment by danipen** [comment]
Ok. AvaloniaEdit.TextMate is coded to make user-edits to the document. So it doesn't handle well those edits that are between:
```csharp
_textEditor.Document.BeginUpdate();
// stuff
_textEditor.Document.EndUpdate();
```

If you remove those calls, the issue dissapear. I'll try to fix it.

**Comment by danipen** [comment]
@guilhermeolisi can you try now changes in #225? That should fix your issue.

**Comment by guilhermeolisi** [comment]
@danipen , i couldn't pull the AvaloniaEdit with your modifications, maybe the problem is the pull permission? 

![image](https://user-images.githubusercontent.com/25594740/162197134-59dcdcf0-c3a6-408e-b0e4-03d698bd525e.png)

**Comment by danipen** [comment]
Can't you do:
`gh pr checkout 225`

Or open it with Github Desktop?

**Comment by guilhermeolisi** [comment]
Well, i tried again in few minutes and i was able to pull.
I tested the AvaloniaEdit.Demo with my button to change the text in code behind e it is working fine now
Thanks for your modifications.
